// @rekog/mcp-nest/dist/mcp/services/mcp-dynamic-registry.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var McpRegistryService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpRegistryService = exports.DYNAMIC_PROMPT_HANDLER_TOKEN = exports.DYNAMIC_RESOURCE_HANDLER_TOKEN = exports.DYNAMIC_TOOL_HANDLER_TOKEN = void 0;
const common_1 = require("@nestjs/common");
const zod_1 = require("zod");
const mcp_registry_discovery_service_1 = require("./mcp-registry-discovery.service");
const mcp_logger_factory_1 = require("../utils/mcp-logger.factory");
exports.DYNAMIC_TOOL_HANDLER_TOKEN = Symbol('DYNAMIC_TOOL_HANDLER');
exports.DYNAMIC_RESOURCE_HANDLER_TOKEN = Symbol('DYNAMIC_RESOURCE_HANDLER');
exports.DYNAMIC_PROMPT_HANDLER_TOKEN = Symbol('DYNAMIC_PROMPT_HANDLER');
const globalHandlers = new Map();
const globalResourceHandlers = new Map();
const globalPromptHandlers = new Map();
let McpRegistryService = McpRegistryService_1 = class McpRegistryService {
    constructor(registry, mcpModuleId, options) {
        this.registry = registry;
        this.mcpModuleId = mcpModuleId;
        this.options = options;
        this.logger = (0, mcp_logger_factory_1.createMcpLogger)(McpRegistryService_1.name, this.options);
        [globalHandlers, globalResourceHandlers, globalPromptHandlers].forEach((store) => {
            if (!store.has(mcpModuleId)) {
                store.set(mcpModuleId, new Map());
            }
        });
    }
    registerTool(definition) {
        this.logger.debug(`Registering dynamic tool: ${definition.name}`);
        const moduleHandlers = globalHandlers.get(this.mcpModuleId);
        moduleHandlers.set(definition.name, definition.handler);
        const parameters = definition.parameters ?? zod_1.z.object({});
        const metadata = {
            name: definition.name,
            description: definition.description,
            parameters,
            outputSchema: definition.outputSchema,
            annotations: definition.annotations,
            _meta: definition._meta,
            isPublic: definition.isPublic,
            requiredScopes: definition.requiredScopes,
            requiredRoles: definition.requiredRoles,
        };
        this.registry.registerDynamicCapability(this.mcpModuleId, {
            type: 'tool',
            metadata,
            providerClass: exports.DYNAMIC_TOOL_HANDLER_TOKEN,
            methodName: definition.name,
        });
    }
    getHandler(toolName) {
        return globalHandlers.get(this.mcpModuleId)?.get(toolName);
    }
    static getHandlerByModuleId(mcpModuleId, toolName) {
        return globalHandlers.get(mcpModuleId)?.get(toolName);
    }
    registerResource(definition) {
        this.logger.debug(`Registering dynamic resource: ${definition.uri}`);
        globalResourceHandlers
            .get(this.mcpModuleId)
            .set(definition.uri, definition.handler);
        const metadata = {
            uri: definition.uri,
            name: definition.name ?? definition.uri,
            description: definition.description,
            mimeType: definition.mimeType,
            _meta: definition._meta,
        };
        this.registry.registerDynamicCapability(this.mcpModuleId, {
            type: 'resource',
            metadata,
            providerClass: exports.DYNAMIC_RESOURCE_HANDLER_TOKEN,
            methodName: definition.uri,
        });
    }
    static getResourceHandlerByModuleId(mcpModuleId, uri) {
        return globalResourceHandlers.get(mcpModuleId)?.get(uri);
    }
    registerPrompt(definition) {
        this.logger.debug(`Registering dynamic prompt: ${definition.name}`);
        globalPromptHandlers
            .get(this.mcpModuleId)
            .set(definition.name, definition.handler);
        const metadata = {
            name: definition.name,
            description: definition.description,
            parameters: definition.parameters,
        };
        this.registry.registerDynamicCapability(this.mcpModuleId, {
            type: 'prompt',
            metadata,
            providerClass: exports.DYNAMIC_PROMPT_HANDLER_TOKEN,
            methodName: definition.name,
        });
    }
    static getPromptHandlerByModuleId(mcpModuleId, name) {
        return globalPromptHandlers.get(mcpModuleId)?.get(name);
    }
    removeTool(name) {
        this.logger.debug(`Removing dynamic tool: ${name}`);
        globalHandlers.get(this.mcpModuleId)?.delete(name);
        this.registry.removeDynamicCapability(this.mcpModuleId, 'tool', name);
    }
    removeResource(uri) {
        this.logger.debug(`Removing dynamic resource: ${uri}`);
        globalResourceHandlers.get(this.mcpModuleId)?.delete(uri);
        this.registry.removeDynamicCapability(this.mcpModuleId, 'resource', uri);
    }
    removePrompt(name) {
        this.logger.debug(`Removing dynamic prompt: ${name}`);
        globalPromptHandlers.get(this.mcpModuleId)?.delete(name);
        this.registry.removeDynamicCapability(this.mcpModuleId, 'prompt', name);
    }
};
exports.McpRegistryService = McpRegistryService;
exports.McpRegistryService = McpRegistryService = McpRegistryService_1 = __decorate([
    (0, common_1.Injectable)(),
    __param(1, (0, common_1.Inject)('MCP_MODULE_ID')),
    __param(2, (0, common_1.Inject)('MCP_OPTIONS')),
    __metadata("design:paramtypes", [mcp_registry_discovery_service_1.McpRegistryDiscoveryService, String, Object])
], McpRegistryService);
