// @rekog/mcp-nest/dist/mcp/services/handlers/mcp-handler.base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpHandlerBase = void 0;
const types_js_1 = require("@modelcontextprotocol/sdk/types.js");
const mcp_logger_factory_1 = require("../../utils/mcp-logger.factory");
const constants_1 = require("@nestjs/common/constants");
const execution_context_host_1 = require("@nestjs/core/helpers/execution-context-host");
class McpHandlerBase {
    constructor(moduleRef, registry, reflector, loggerContext, options) {
        this.moduleRef = moduleRef;
        this.registry = registry;
        this.reflector = reflector;
        this.logger = (0, mcp_logger_factory_1.createMcpLogger)(loggerContext, options);
    }
    createContext(mcpServer, mcpRequest) {
        if (mcpServer.server.transport.sessionId === undefined) {
            return this.createStatelessContext(mcpServer, mcpRequest);
        }
        const progressToken = mcpRequest.params?._meta?.progressToken;
        return {
            reportProgress: async (progress) => {
                if (progressToken) {
                    await mcpServer.server.notification({
                        method: 'notifications/progress',
                        params: {
                            ...progress,
                            progressToken,
                        },
                    });
                }
            },
            log: {
                debug: (message, context) => {
                    void mcpServer.server.sendLoggingMessage({
                        level: 'debug',
                        data: { message, context },
                    });
                },
                error: (message, context) => {
                    void mcpServer.server.sendLoggingMessage({
                        level: 'error',
                        data: { message, context },
                    });
                },
                info: (message, context) => {
                    void mcpServer.server.sendLoggingMessage({
                        level: 'info',
                        data: { message, context },
                    });
                },
                warn: (message, context) => {
                    void mcpServer.server.sendLoggingMessage({
                        level: 'warning',
                        data: { message, context },
                    });
                },
            },
            mcpServer,
            mcpRequest,
        };
    }
    createStatelessContext(mcpServer, mcpRequest) {
        const warn = (fn) => {
            this.logger.warn(`Stateless context: '${fn}' is not supported.`);
        };
        return {
            reportProgress: async (_progress) => {
                warn('reportProgress not supported in stateless');
            },
            log: {
                debug: (_message, _data) => {
                    warn('server report logging not supported in stateless');
                },
                error: (_message, _data) => {
                    warn('server report logging not supported in stateless');
                },
                info: (_message, _data) => {
                    warn('server report logging not supported in stateless');
                },
                warn: (_message, _data) => {
                    warn('server report logging not supported in stateless');
                },
            },
            mcpServer,
            mcpRequest,
        };
    }
    createErrorResponse(errorText) {
        throw new types_js_1.McpError(types_js_1.ErrorCode.InternalError, errorText);
    }
    handleError(error, capabilityInfo, httpRequest) {
        this.logger.error(error);
        if (error instanceof types_js_1.McpError) {
            throw error;
        }
        const clazz = capabilityInfo.providerClass;
        const method = clazz.prototype[capabilityInfo.methodName];
        const methodFilters = this.reflector.get(constants_1.EXCEPTION_FILTERS_METADATA, method) ?? [];
        const classFilters = this.reflector.get(constants_1.EXCEPTION_FILTERS_METADATA, clazz) ?? [];
        const allFilters = [...methodFilters, ...classFilters];
        for (const FilterClass of allFilters) {
            if (this.isExceptionFiltered(error, FilterClass)) {
                const filterInstance = new FilterClass();
                const host = new execution_context_host_1.ExecutionContextHost([httpRequest.raw], capabilityInfo.providerClass, method);
                host.setType('http');
                const result = filterInstance.catch(error, host);
                const text = typeof result === 'string' ? result : JSON.stringify(result);
                return this.createErrorResponse(text);
            }
        }
        return this.createErrorResponse(error.message);
    }
    getExceptionTypes(filter) {
        return (this.reflector.get(constants_1.FILTER_CATCH_EXCEPTIONS, filter) ?? []);
    }
    isExceptionFiltered(error, filter) {
        const exceptionTypes = this.getExceptionTypes(filter);
        if (exceptionTypes.length === 0) {
            return true;
        }
        return exceptionTypes.some((type) => error instanceof type);
    }
}
exports.McpHandlerBase = McpHandlerBase;
